### ISSUE TITLE
[Bug]: No protection against stack overflows.

### ISSUE THREAD

**Issue by MikaelMayer** [dafny-lang/dafny#3046]
### Dafny version

3.9.1

### Code to produce this issue

```dafny
datatype Oddity = Even | Odd

function method Helper(i: nat, s: string): Oddity
  decreases |s|-i
  requires i <= |s|
{
  if i == |s| then Even
  else match Helper(i + 1, s) {
    case Even => Odd
    case Odd => Even
  }
}

method Main() {
  var longString := seq(5000, i => 'h');
  print Helper(0, longString);
}
```

### Command to run and results

```code
dafny stackoverflow.dfy /compile:3


   at _module.__default.Helper(System.Numerics.BigInteger, Dafny.ISequence`1<Char>)
   at _module.__default.Helper(System.Numerics.BigInteger, Dafny.ISequence`1<Char>)
   ... 1000 times
   at _module.__default.Helper(System.Numerics.BigInteger, Dafny.ISequence`1<Char>)
   at _module.__default.Helper(System.Numerics.BigInteger, Dafny.ISequence`1<Char>)
```

### What happened?

I was expecting that Dafny would warn me against potential stack overflows so that I would not encounter this issue at run-time.

For example, I would love that Dafny adds the following error message:

```
print Helper(0, longString);
      ^^^^^^^^^^^
Cannot prove that this won't stack overflow.
```

The C# compiler could be responsible for adding this assertion at verification time, since it's concerned by stack overflows.
The way it could prove this is that, it could consider the max stack size M, and look at the "decreases" clause, 
and attempt to prove
`assert {:error "Cannot prove that this won't stack overflow"} |s| - i < M`

For a datatype, it could create an implicit boogie "height" option around it, and would try to prove that
```
`assert {:error "Cannot prove that this won't stack overflow"} height(tree) < M`
```

It might get more complex than that wish mutually recursive functions, but I was pretty annoyed that I did a lot of effort writing something in Dafny just to discover it crashed at run-time.

### What type of Operating System are you seeing the problem on?

Windows

**Comment by cpitclaudel** [comment]
What you're seeing is a well-known ~bug~ ~limitation~ design choice in the dotnet ecosystem.  

The .Net framework sets a low stack size limit (~1MB), which unfortunately breaks all sorts of reasonable programs.  They do this, if I recall correctly, because the .net GC handles the stack differently from the heap, and large stacks lead to performance issues. https://stackoverflow.com/questions/7947446/why-does-f-impose-a-low-limit-on-stack-size has a discussion of this issue in the context of F#.

There is nothing fundamental about stack overflows.  In fact, many languages, including some of the ones we support, handle your program without issue:

```
> dafny run --target go test.dfy

Dafny program verifier finished with 2 verified, 0 errors
Oddity.Even
```

You can easily fix the issue in C# by increasing the stack size; it's annoying that C# doesn't respect system limits (at least as far as I can tell — I'd love to be proven wrong!): running with `ulimit -s unlimited` still leads to overflows.

We could mitigate the issue by not using recursion when compiling to C# (and to other languages that suffer from the same issue).  It would be a non-trivial amount of work, but it would be very useful.

---

I realize that the above doesn't directly address the "proof" aspect.  That's because I doubt that, in it's current state, Dafny can say anything useful about the stack (and I question whether "the stack" is even a useful abstraction at the level we're dealing with).

This is because, unfortunately, *all* functions can overflow the stack (even non-recursive ones: a deep-enough chain of non-recursive calls can overflow the stack!).

The problem is that "will not stack overflow" is not a local property: you need to state which context you're calling the function in.   In other words, to stick with the message you suggested, we'd have to put "Cannot prove that this won't stack overflow." on *all* functions, because there's nothing that prevents you from calling the function in a context where almost no stack space remains.

Now, what we *could* imagine doing is adding resource annotations to Dafny, and over-approximating stack usage for non-recursive functions (we don't have a precise stack model at the moment, so we don't know how much stack a given function will use, but we could work on that).  It would be a lot of work.

There are two problems with this:

1. It doesn't help with recursive functions, because their stack bounds are dependent on their input size.
2. We still have to check, when the program starts, that we have enough stack, and report a stack overflow before the program start if we don't (we do need the check, because different systems can allocate different amount of stack space to our process, and we may not have control over that).

We can solve these two issues by getting rid of the stack; either explicitly, essentially by trampolining all function calls, or implicitly, by dynamically allocating our own stack space on the heap (IIRC the `nim` programming language has support for this).

But once we do this, do we really need to track stack space usage?  Probably not.

Related issue: https://github.com/dafny-lang/dafny/issues/2859

**Comment by MikaelMayer** [comment]
> We can solve these two issues by getting rid of the stack; either explicitly, essentially by trampolining all function calls, or implicitly, by dynamically allocating our own stack space on the heap (IIRC the nim programming language has support for this).

Doing our own compilation (maintaining our own stack in the heap) would mean even less interoperability with other languages unfortunately.
You nailed most of the objections I would have had in mind myself about my proposal.

But now I can't stop thinking, we still have a soundness issue. Wouldn't it be great if Dafny was aware of these potential issues and prevented them from happening before execution? In the example I gave, once I found the stack overflow, it was not hard for me to transform the code to a tail-recursive version.
But it would have been much nicer if I had to tell Dafny at compile-time that there are no stack overflows.
So let me try to answer our common objections.

**The stack size depends on the language**
That's fine, compilers are plugins now, so we could always add more verification items inside them. It's acceptable that verification would not produce the same results depending on the target. After all, the runtimes don't all offer the same restrictions.

**all functions can overflow the stack (even non-recursive ones: a deep-enough chain of non-recursive calls can overflow the stack!).**
If we consider a Dafny program as the entry point, then we could display an error at the entry point of course. The error would not be display on lower non-recursive functions, but on the "ancestor" which cannot guarantee that his descendant call stack is less than a certain threshold.
If Dafny was meant to be a library to be called, we could add options to even further restrict the amount of stack available to offer even better guarantees.

**we'd have to put "Cannot prove that this won't stack overflow." on all functions, because there's nothing that prevents you from calling the function in a context where almost no stack space remains.**
A non-recursive function (or tail-recursive function) that does not call any other function would fail to prove it does not overflow if the stack size it requires for a single call exceeds the global stack size 
Any others would need to prove that their own call + variables + descendant calls  don't exceed the global stack size. See the previous point. Errors would be thrown on outer functions (during verification), not inner functions (like it's the case during execution)

**we don't have a precise stack model at the moment, so we don't know how much stack a given function will use, but we could work on that**
We could definitely work on that. Since this kind of verification happens from the compiler plugin itself, the compiler is aware of the language and of the compiled types, so it can know what is the necessary stack (whether it's a number of calls or a memory limitation)

**t would be a lot of work.**
Most soundness issues require a lot of work and design :-)

**It doesn't help with recursive functions, because their stack bounds are dependent on their input size.**
Yes it does help with recursive functions. It is true that their stack bounds depend on their input size, so that's where we need to have a clever design than "decreases". For example, we could restrict "decreases" clauses of non-ghost functions and methods to be integers expressions only, or integers expressions followed by any number of ", 1". If a decreases expression is N(), 1, 1, 1; we know that the maximum stack if can consume is given by the maximum stack any function going outside of the clique can consume + 4*N()*(the maximum stack a function in the clique can use)

**We still have to check, when the program starts, that we have enough stack**
If Dafny is used as a stand-alone application in dotnet, then we know the stack size: it's 2Mb.
If Dafny is used as a library (meaning we provide much less guarantees anyway), then we could at least verify that the stack size is less than 1Mb statically.
In any case, the purpose of verifying the stack is to discover potential stack overflows at compile-time, which can be typical in a recursive setting which is not tail recursive
